### PR TITLE
feat(algorithms): add bucket sort for array of numbers

### PR DESCRIPTION
**Pull request by jp-tran**

### CODE REVIEW

File: test/algorithms/sort/bucket-sort.test.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    bucketSort(array, undefined, 1)
+    expect(array).toEqual(soln)
+  })
+
+  it('raises error when number of specified buckets <= 0', () => {
+    const array = [1, 2, 3]
+    bucketSort(array, 0, 1)
+    expect(console.error).toHaveBeenCalledTimes(1)
+  })
+})
```
Comment (jeffzh4ng) [inline test/algorithms/sort/bucket-sort.test.ts:39]: We need to test the case where there's negative and positive numbers in the array.

**Review by jeffzh4ng — CHANGES_REQUESTED** [review]

File: test/algorithms/sort/bucket-sort.test.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    const array = [-5, -3, -4, -1, -2]
+    const soln = [...array].sort((a, b) => a - b)
+    bucketSort(array, undefined, 1)
+    expect(array).toEqual(soln)
+  })
+
+  it('correctly sorts when both posBuckets and negBuckets are specified', () => {
+    const array = [-5, -3, 7, -4, -1, -2, 9, 0]
+    const soln = [...array].sort((a, b) => a - b)
+    bucketSort(array, 5, 3)
```
Comment (jeffzh4ng) [inline test/algorithms/sort/bucket-sort.test.ts:37]: wait, how is this test passing? shouldn't it be `bucketSort(array, 3, 5)`?

Comment (jp-tran) [reply test/algorithms/sort/bucket-sort.test.ts:37]: The number of buckets specified shouldn't affect the algorithm's ability to sort correctly as long as there is at least one bucket to place elements into. 

If there are fewer buckets than array elements, some buckets will end up with more than one element. In this case, I think the algorithm's runtime would approach O(n^2) as the ratio of element : bucket approaches the length of the array we're sorting (our worst case time complexity). If there are more buckets than elements, some buckets will end up empty. In this second case, the run time would be O(n + k), it's just that k (the number of buckets) isn't optimal.

Since I give users the option to specify the optimal number of buckets, I think we should make sure that it still sorts correctly if they choose too many or too few buckets.

**Review by jeffzh4ng — APPROVED** [review]
